src/jit.c

```diff
@@ -0,0 +1,1550 @@
+                                    int dst,
+                                    int32_t offset,
+                                    int32_t imm)
+{
+#if defined(__x86_64__)
+    if (size == S16)
+        emit1(state, 0x66); /* 16-bit override */
+    emit1(state, size == S8 ? 0xc6 : 0xc7);
+    emit_modrm_and_displacement(state, 0, dst, offset);
+    if (size == S32)
+        emit4(state, imm);
+    else if (size == S16)
+        emit2(state, imm);
+    else if (size == S8)
+        emit1(state, imm);
+#elif defined(__aarch64__)
+    emit_load_imm(state, R10, imm);
+    emit_store(state, size, R10, dst, offset);
+#endif
+}
+
+static inline void emit_jmp(struct jit_state *state, uint32_t target_pc)
+{
+#if defined(__x86_64__)
+    emit1(state, 0xe9);
+    emit_jump_target_address(state, target_pc);
+#elif defined(__aarch64__)
+    struct jump *jump = &state->jumps[state->num_jumps++];
+    jump->offset_loc = state->offset;
+    jump->target_pc = target_pc;
+    emit_instruction(state, UBR_B);
+#endif
+}
+
+static inline void emit_call(struct jit_state *state, intptr_t target)
+{
+#if defined(__x86_64__)
+    emit_load_imm(state, RAX, target);
+    /* callq *%rax */
+    emit1(state, 0xff);
+    /* ModR/M byte: b11010000b = xd0, rax is register 0 */
+    emit1(state, 0xd0);
+#elif defined(__aarch64__)
+    uint32_t stack_movement = align_to(8, 16);
+    emit_addsub_immediate(state, true, AS_SUB, SP, SP, stack_movement);
+    emit_loadstore_immediate(state, LS_STRX, R30, SP, 0);
+
+    emit_movewide_immediate(state, true, temp_imm_reg, target);
+    emit_unconditionalbranch_register(state, BR_BLR, temp_imm_reg);
+
+    /* On exit need to move result from r0 to whichever register we've mapped
+     * EBPF r0 to.  */
+    int dest = map_register(0);
+    if (dest != R0) {
+        emit_logical_register(state, true, LOG_ORR, dest, RZ, R0);
+    }
+
+    emit_loadstore_immediate(state, LS_LDRX, R30, SP, 0);
+    emit_addsub_immediate(state, true, AS_ADD, SP, SP, stack_movement);
+#endif
+}
+
+static inline void emit_exit(struct jit_state *state)
+{
+#if defined(__x86_64__)
+    emit1(state, 0xe9);
+    emit_jump_target_offset(state, state->offset, state->exit_loc);
+    emit4(state, 0);
+#elif defined(__aarch64__)
+    emit_jmp(state, TARGET_PC_EXIT);
+#endif
+}
+
+#if RV32_HAS(EXT_M)
+static void muldivmod(struct jit_state *state,
+                      uint8_t opcode,
+                      int src,
+                      int dst,
+                      int32_t imm UNUSED)
+{
+#if defined(__x86_64__)
+    bool mul = (opcode & JIT_ALU_OP_MASK) == (JIT_OP_MUL_IMM & JIT_ALU_OP_MASK);
+    bool div = (opcode & JIT_ALU_OP_MASK) == (JIT_OP_DIV_IMM & JIT_ALU_OP_MASK);
+    bool mod = (opcode & JIT_ALU_OP_MASK) == (JIT_OP_MOD_IMM & JIT_ALU_OP_MASK);
+    bool is64 = (opcode & JIT_CLS_MASK) == JIT_CLS_ALU64;
+    bool reg = (opcode & JIT_SRC_REG) == JIT_SRC_REG;
+
+    /* Short circuit for imm == 0 */
+    if (!reg && imm == 0) {
+        assert(NULL);
```

Comment:
Are you ready to close #297 ?

Reply:
Not yet. I will close this issue by another PR.

Reply:
> Not yet. I will close this issue by another PR.

Add proper `FIXME` or `TODO` comment to reminder.